toolchain/base/value_store.h

```diff
@@ -113,9 +175,50 @@ class ValueStore
   }
 
  private:
+  // Poison the entire contents of the value store. This is used to detect cases
+  // where references to elements in a value store are used across calls that
+  // might modify the store.
+  [[gnu::always_inline]] auto PoisonAll() const -> void {
+#if CARBON_POISON_VALUE_STORES
+    if (!all_poisoned_) {
+      __asan_poison_memory_region(values_.data(),
+                                  values_.size() * sizeof(values_[0]));
+      all_poisoned_ = true;
+    }
+#endif
+  }
+  // Unpoison the entire contents of the value store.
+  [[gnu::always_inline]] auto UnpoisonAll() const -> void {
+#if CARBON_POISON_VALUE_STORES
+    __asan_unpoison_memory_region(values_.data(),
+                                  values_.size() * sizeof(values_[0]));
+    all_poisoned_ = false;
+#endif
+  }
+  // Poison a single element.
+  [[gnu::always_inline]] auto PoisonElement([[maybe_unused]] int element) const
+      -> void {
+#if CARBON_POISON_VALUE_STORES
+    __asan_unpoison_memory_region(values_.data() + element, sizeof(values_[0]));
+#endif
+  }
+  // Unpoison a single element.
+  [[gnu::always_inline]] auto UnpoisonElement(
+      [[maybe_unused]] int element) const -> void {
+#if CARBON_POISON_VALUE_STORES
+    __asan_unpoison_memory_region(values_.data() + element, sizeof(values_[0]));
+    all_poisoned_ = false;
+#endif
+  }
+
   // Set inline size to 0 because these will typically be too large for the
   // stack, while this does make File smaller.
   llvm::SmallVector<std::decay_t<ValueType>, 0> values_;
+
+#if CARBON_POISON_VALUE_STORES
+  // Whether the vector is currently fully poisoned.
+  mutable bool all_poisoned_ = true;
+#endif
```

Comment:
A little sad about the need for this -- I assume the issue is that poisoning repeatedly is slow?

Reply:
That's my concern, yes -- we poison every time we import anything, and poisoning is linear in the size of the value stores, so I'm trying to keep the full poisonings to a minimum. But it's not possible to measure the performance impact yet because the tests all crash with this enabled.